## Week 8 Hw Python Part 1

Part 1: Canids continued                                       
1a. Using the python code we generated in Thursday's class as a start do the following:                                                     

    1-Remove quotes from species names                          

In [12]:
## open the file

infile = "/home/eeb177-student/Desktop/eeb-177/class-assignments/canids/pbdbformatted.csv"
with open(infile, 'r') as ff:
    canids_recs = ff.readlines()

In [24]:
# use default dict
from collections import defaultdict
species_ranges = defaultdict(list)


# make a dict
for line in canids_recs:
    species = line.split(",")[5][1:-1]
    #types = line.split(",")[6][1:-1]
    minage = line.split(",")[10][1:-1]
    maxage = line.split(",")[11][1:-1]
    mean_age = (float(minage) + float(maxage))/2
    #print(mean_age)
    #species_ranges[species]
    
    species_ranges[species].append(mean_age) # add speices as keyes and mean ages as elements to the list of ages
    





    2-Sort the output so that species are listed alphabetically.
   

In [14]:
sorted_list= sorted(species_ranges.keys())
for key in sorted_list:
    ages = species_ranges[key]
    minage = min(ages)
    maxage = max(ages)
    genus = key.split(" ")[0]
    species = key
    outline= "{},{},{},{}\n".format(genus, species, minage, maxage)
    print(outline)

Aelurodon,Aelurodon,8.470500000000001,14.785

Aelurodon,Aelurodon asthenostylus,14.785,18.2

Aelurodon,Aelurodon ferox,9.25,14.785

Aelurodon,Aelurodon mcgrewi,14.1815,14.785

Aelurodon,Aelurodon montanensis,14.785,14.785

Aelurodon,Aelurodon stirtoni,11.95,14.785

Aelurodon,Aelurodon taxoides,7.6000000000000005,14.1815

Aelurodon,Aelurodon wheelerianus,18.2,18.2

Alopex,Alopex,2.6723500000000002,2.6723500000000002

Alopex,Alopex lagopus,0.00585,1.05

Archaeocyon,Archaeocyon falkenbachi,25.615000000000002,25.615000000000002

Archaeocyon,Archaeocyon leptodus,25.615000000000002,32.05

Archaeocyon,Archaeocyon pavidus,25.615000000000002,32.05

Borophaginae,Borophaginae,7.6000000000000005,22.615000000000002

Borophagus,Borophagus,3.35,14.1815

Borophagus,Borophagus diversidens,3.35,3.35

Borophagus,Borophagus dudleyi,4.4665,4.4665

Borophagus,Borophagus hilli,3.35,14.1815

Borophagus,Borophagus littoralis,11.95,14.1815

Borophagus,Borophagus orc,4.4665,7.6000000000000005

Borophagus,Boropha

    Write a function that does the following:                     
        takes a file name as argument                                  
        opens the file that is formatted like the cleaned version of the canid file we made from the shell in class.                            
        extracts the species name for each species with a latin binomial.
        ignores records that have only the genus recorded.             
        populates a dictionary where the species are keys and the values are the ages of fossils known for each species                           
        returns this dictionary                                          
    

In [26]:
def species_dict_2(reader):
    from collections import defaultdict

import csv
#output = open("homin-ranges.csv", "w")
with open("/home/eeb177-student/Desktop/eeb-177/class-assignments/canids/pbdbformatted.csv", newline='') as infh:
    reader = csv.reader(infh)
species_ranges = defaultdict(list)

def extract_func(species):
    #species_dict = {}
    for row in reader:
        if row[6] == "species": #this will only get species and not genus
            species = line.split(",")[5][1:-1]
            #types = line.split(",")[6][1:-1]
            minage = line.split(",")[10][1:-1]
            maxage = line.split(",")[11][1:-1]
            mean_age = (float(minage) + float(maxage))/2
            #print(mean_age)
            #species_ranges[species]
    
            species_ranges[species].append(mean_age)
          
            
            
            print(row[5] + "," + row[10] + "," + row[11])
            
        return(species_ranges)
            



Write a second function that does the following:                          

    takes a dictionary and output file name as arguments               
    opens a file for writing given the file name in the argument.         
    determines the minimum and maximum fossil age for each species in the passed dictionary                                                         
    writes to the outfile a csv with the following fields: genus, species, minage, maxage                                                          

In [27]:
output = open("canid-ranges2.csv", "w")

#get oldest and youngest
#species_ranges.values()
for key in species_ranges.keys():
    ages = species_ranges[key]
    minage = min(ages)
    maxage = max(ages)
    genus = key.split(" ")[0]
    species = key
    outline= "{},{},{},{}\n".format(genus, species, minage, maxage)
    output.write(outline)
    #print(genus, key, minage, maxage)
    #output.write(genus + "," + key + ",")
    #print(outline) 
    
output.close()

Write a third function that finds the species with the longest fossil range given a dictionary of species ranges             

In [35]:
def longest_fossil(species_ranges):
    fossil_time= sorted(species_ranges.keys())
    longest_name = ""
    longest_foss = 0
    for line in fossil_time:
        ages = species_ranges[key]
        minage = min(ages)
        maxage = max(ages)
        time = (maxage - minage)
        if time > longest_foss:
            longest_foss = time
            longest_name = species
    fossil_range = longest_name + " " + str(longest_foss)
    return fossil_range

In [43]:
import csv
#output = open("homin-ranges.csv", "w")
with open("/home/eeb177-student/Desktop/eeb-177/class-assignments/canids/pbdbformatted.csv", newline='') as infh:
    reader = csv.reader(infh)
    
#def extract_func(species):
    for row in reader:
        if row[6] == "species":
            genus = row[5].split(" ")[0]
            print(genus + "," + row[5] + "," + row[11] + "," + row[10])

Cynodictis,Cynodictis lacustris,33.9,37.2
Neovulpavus,Neovulpavus washakius,40.4,46.2
Hesperocyon,Hesperocyon gregarius,37.2,40.4
Hesperocyon,Hesperocyon gregarius,33.9,37.2
Hesperocyon,Hesperocyon gregarius,33.9,37.2
Hesperocyon,Hesperocyon gregarius,33.9,37.2
Hesperocyon,Hesperocyon gregarius,33.9,37.2
Hesperocyon,Hesperocyon gregarius,33.9,37.2
Hesperocyon,Hesperocyon gregarius,33.9,37.2
Hesperocyon,Hesperocyon gregarius,33.9,37.2
Hesperocyon,Hesperocyon gregarius,33.9,37.2
Hesperocyon,Hesperocyon gregarius,33.9,37.2
Hesperocyon,Hesperocyon gregarius,33.9,37.2
Hesperocyon,Hesperocyon gregarius,33.9,37.2
Hesperocyon,Hesperocyon gregarius,33.9,37.2
Hesperocyon,Hesperocyon gregarius,33.9,37.2
Hesperocyon,Hesperocyon gregarius,33.9,37.2
Hesperocyon,Hesperocyon gregarius,33.9,37.2
Hesperocyon,Hesperocyon gregarius,33.9,37.2
Hesperocyon,Hesperocyon gregarius,33.9,37.2
Hesperocyon,Hesperocyon gregarius,33.9,37.2
Hesperocyon,Hesperocyon gregarius,33.9,37.2
Hesperocyon,Hesperocyon gregarius,

## Visulaizing in R

   library(ggplot2)

    canids <- read.csv("~/Desktop/eeb-177/homework/canidsoutput.csv", header = F, as.is = T)
    names(canids) <- c("genus", "species", "minage", "maxage")
    head(canids)

    ##         genus               species minage maxage
    ## 1  Cynodictis  Cynodictis lacustris   33.9   37.2
    ## 2 Neovulpavus Neovulpavus washakius   40.4   46.2
    ## 3 Hesperocyon Hesperocyon gregarius   37.2   40.4
    ## 4 Hesperocyon Hesperocyon gregarius   33.9   37.2
    ## 5 Hesperocyon Hesperocyon gregarius   33.9   37.2
    ## 6 Hesperocyon Hesperocyon gregarius   33.9   37.2

    canid_occ <- ggplot(canids, aes( species, ymin = maxage, ymax=minage, colour = genus))
    canid_occ <- canid_occ + geom_linerange()
    canid_occ

![](week8-hw-part1-Visulaizing-canids_files/figure-markdown_strict/unnamed-chunk-3-1.png)

    canid_occ <- canid_occ + theme(legend.position="none")
    canid_occ

![](week8-hw-part1-Visulaizing-canids_files/figure-markdown_strict/unnamed-chunk-4-1.png)

    canid_occ <- canid_occ + coord_flip()
    canid_occ

![](week8-hw-part1-Visulaizing-canids_files/figure-markdown_strict/unnamed-chunk-5-1.png)

    canid_occ <- canid_occ +  theme(axis.text.y = element_text(size=3))
    canid_occ

![](week8-hw-part1-Visulaizing-canids_files/figure-markdown_strict/unnamed-chunk-6-1.png)

    canid_occ <- canid_occ + theme(axis.ticks.y=element_blank())
    canid_occ

![](week8-hw-part1-Visulaizing-canids_files/figure-markdown_strict/unnamed-chunk-7-1.png)

    canid_occ <- canid_occ + scale_y_continuous(limits=c(0, 40), expand = c(0, 0), breaks=c(0, 10, 20, 30, 40))
    canid_occ

    ## Warning: Removed 3 rows containing missing values (geom_linerange).

![](week8-hw-part1-Visulaizing-canids_files/figure-markdown_strict/unnamed-chunk-8-1.png)

    canid_occ <- canid_occ + labs(title = "Canid Fossil Occurrences", x = "Species", y = "Ma ago") + theme(plot.title = element_text(hjust = 0.5, size=22, face = "bold"), axis.title =element_text(size=20))
    canid_occ

    ## Warning: Removed 3 rows containing missing values (geom_linerange).

![](week8-hw-part1-Visulaizing-canids_files/figure-markdown_strict/unnamed-chunk-9-1.png)


Repeat parts 1a and 1b with a new family or order from the PBDB. You may use the taxonomic group you have chosen for your final project or another group. Make sure that whatever group you decide to use has at least 100 unique species. 

In [46]:
## open the file

infile = "/home/eeb177-student/Desktop/eeb-177/homework/catarrhiniformatted.csv"
with open(infile, 'r') as ff:
    canids_recs = ff.readlines()

In [49]:
# use default dict
from collections import defaultdict
species_ranges = defaultdict(list)


# make a dict
for line in canids_recs:
    species = line.split(",")[5][1:-1]
    #types = line.split(",")[6][1:-1]
    minage = line.split(",")[15][1:-1]
    maxage = line.split(",")[14][1:-1]
    #mean_age = (float(minage) + float(maxage))/2
    #print(mean_age)
    #species_ranges[species]
    
    species_ranges[species].append(mean_age) # add speices as keyes and mean ages as elements to the list of ages
    




In [50]:
sorted_list= sorted(species_ranges.keys())
for key in sorted_list:
    ages = species_ranges[key]
    minage = min(ages)
    maxage = max(ages)
    genus = key.split(" ")[0]
    species = key
    outline= "{},{},{},{}\n".format(genus, species, minage, maxage)
    print(outline)

Aegyptopithecus,Aegyptopithecus zeuxis,3.9605,3.9605

Afropithecus,Afropithecus turkanensis,3.9605,3.9605

Anapithecus,Anapithecus hernyaki,3.9605,3.9605

Ardipithecus,Ardipithecus kadabba,3.9605,3.9605

Australopithecus,Australopithecus afarensis,3.9605,3.9605

Australopithecus,Australopithecus africanus,3.9605,3.9605

Australopithecus,Australopithecus anamensis,3.9605,3.9605

Australopithecus,Australopithecus boisei,3.9605,3.9605

Australopithecus,Australopithecus cf. afarensis,3.9605,3.9605

Australopithecus,Australopithecus cf. africanus,3.9605,3.9605

Australopithecus,Australopithecus cf. boisei,3.9605,3.9605

Australopithecus,Australopithecus garhi,3.9605,3.9605

Australopithecus,Australopithecus n. sp. ramidus,3.9605,3.9605

Australopithecus,Australopithecus n. sp. sediba,3.9605,3.9605

Australopithecus,Australopithecus robustus,3.9605,3.9605

Australopithecus,Australopithecus sp.,3.9605,3.9605

Cercopithecoidea,Cercopithecoidea indet.,3.9605,3.9605

Dryopithecus,Dryopithecus ? 

In [55]:
def species_dict_2(reader):
    from collections import defaultdict

import csv
#output = open("homin-ranges.csv", "w")
with open("/home/eeb177-student/Desktop/eeb-177/homework/catarrhiniformatted.csv", newline='') as infh:
    reader = csv.reader(infh)
species_ranges = defaultdict(list)

def extract_func(species):
    #species_dict = {}
    for row in reader:
        if row[6] == "species": #this will only get species and not genus
            species = line.split(",")[5][1:-1]
            #types = line.split(",")[6][1:-1]
            minage = line.split(",")[10][1:-1]
            maxage = line.split(",")[11][1:-1]
            mean_age = (float(minage) + float(maxage))/2
            #print(mean_age)
            #species_ranges[species]
    
            species_ranges[species].append(mean_age)
          
            
            
            print(row[5] + "," + row[10] + "," + row[11])
            
        return(species_ranges)
            

In [56]:
output = open("catarr-ranges2.csv", "w")

#get oldest and youngest
#species_ranges.values()
for key in species_ranges.keys():
    ages = species_ranges[key]
    minage = min(ages)
    maxage = max(ages)
    genus = key.split(" ")[0]
    species = key
    outline= "{},{},{},{}\n".format(genus, species, minage, maxage)
    output.write(outline)
    #print(genus, key, minage, maxage)
    #output.write(genus + "," + key + ",")
    #print(outline) 
    
output.close()

In [57]:
def longest_fossil(species_ranges):
    fossil_time= sorted(species_ranges.keys())
    longest_name = ""
    longest_foss = 0
    for line in fossil_time:
        ages = species_ranges[key]
        minage = min(ages)
        maxage = max(ages)
        time = (maxage - minage)
        if time > longest_foss:
            longest_foss = time
            longest_name = species
    fossil_range = longest_name + " " + str(longest_foss)
    return fossil_range

In [58]:
import csv
#output = open("homin-ranges.csv", "w")
with open("/home/eeb177-student/Desktop/eeb-177/homework/catarrhiniformatted.csv", newline='') as infh:
    reader = csv.reader(infh)
    
#def extract_func(species):
    for row in reader:
        if row[6] == "species":
            genus = row[5].split(" ")[0]
            print(genus + "," + row[5] + "," + row[11] + "," + row[10])

Lufengpithecus,Lufengpithecus lufengensis,89270,genus
Laccopithecus,Laccopithecus robustus,58366,species
Sivapithecus,Sivapithecus indicus,133676,species
Pongo,Pongo pygmaeus,232338,species
Homo,Homo sapiens,83088,species
Homo,Homo sapiens,83088,species
Homo,Homo erectus,83084,species
Australopithecus,Australopithecus cf. africanus,83066,species
Homo,Homo sapiens,83088,species
Proconsul,Proconsul cf. africanus,157732,species
Homo,Homo cf. erectus,83084,species
Homo,Homo cf. erectus,83084,species
Homo,Homo cf. erectus,83084,species
Homo,Homo aff. erectus,83084,species
Homo,Homo cf. erectus,83084,species
Australopithecus,Australopithecus boisei,83068,species
Homo,Homo cf. erectus,83084,species
Australopithecus,Australopithecus boisei,83068,species
Australopithecus,Australopithecus cf. boisei,83068,species
Australopithecus,Australopithecus boisei,83068,species
Homo,Homo sapiens,83088,species
Homo,Homo sapiens,83088,species
Homo,Homo sapiens,83088,species
Homo,Homo sapiens,83088,species
Ho

   library(ggplot2)
    catarr <- read.csv("~/Desktop/eeb-177/eeb-174-final-project/catarrhinioutput.csv", header = F, as.is = T)
    names(catarr) <- c("genus", "species", "minage", "maxage")
    head(catarr)

    ##            genus                    species  minage maxage
    ## 1 Lufengpithecus Lufengpithecus lufengensis  8.7000 5.3330
    ## 2  Laccopithecus     Laccopithecus robustus  8.7000 5.3330
    ## 3   Sivapithecus       Sivapithecus indicus 11.6080 5.3330
    ## 4          Pongo             Pongo pygmaeus  0.7810 0.0117
    ## 5           Homo               Homo sapiens  0.0117 0.0000
    ## 6           Homo               Homo sapiens  0.1260 0.0117

    catarr_occ <- ggplot(catarr, aes( species, ymin = maxage, ymax=minage, colour = genus))
    catarr_occ <- catarr_occ + geom_linerange()
    catarr_occ

![](catarr_occurence_in_R_files/figure-markdown_strict/unnamed-chunk-2-1.png)

    catarr_occ <- catarr_occ + theme(legend.position="none")
    catarr_occ

![](catarr_occurence_in_R_files/figure-markdown_strict/unnamed-chunk-3-1.png)

    catarr_occ <- catarr_occ + coord_flip()
    catarr_occ

![](catarr_occurence_in_R_files/figure-markdown_strict/unnamed-chunk-4-1.png)

    catarr_occ <- catarr_occ + theme(legend.position="none")
    catarr_occ

![](catarr_occurence_in_R_files/figure-markdown_strict/unnamed-chunk-5-1.png)

    catarr_occ <- catarr_occ +  theme(axis.text.y = element_text(size=3))
    catarr_occ

![](catarr_occurence_in_R_files/figure-markdown_strict/unnamed-chunk-6-1.png)

    catarr_occ <- catarr_occ + theme(axis.ticks.y=element_blank())
    catarr_occ

![](catarr_occurence_in_R_files/figure-markdown_strict/unnamed-chunk-7-1.png)

    catarr_occ <- catarr_occ + scale_y_continuous(limits=c(0, 30), expand = c(0, 0), breaks=c(0, 10, 20, 30, 40))
    catarr_occ

    ## Warning: Removed 11 rows containing missing values (geom_linerange).

![](catarr_occurence_in_R_files/figure-markdown_strict/unnamed-chunk-8-1.png)

    catarr_occ <- catarr_occ + labs(title = "Catarrhini Fossil Occurrences", x = "Species", y = "Ma ago") + theme(plot.title = element_text(hjust = 0.5, size=22, face = "bold"), axis.title =element_text(size=20))
    catarr_occ

    ## Warning: Removed 11 rows containing missing values (geom_linerange).

![](catarr_occurence_in_R_files/figure-markdown_strict/unnamed-chunk-9-1.png)
